File: app/src/main/java/com/developersian/util/Utilities/FileUtils.java
package com.developersian.util.Utilities;
import android.annotation.*;
import android.content.*;
import android.database.*;
import android.net.*;
import android.os.*;
import android.provider.*;
public class FileUtils{
	public static String getRealPath(Context context,Uri fileUri){
		String realPath;
		realPath=FileUtils.getRealPathFromURI_API19(context,fileUri);
		return realPath;
	}
	@SuppressLint("NewApi") public static String getRealPathFromURI_API11to18(Context context,Uri contentUri){
		String[] proj={MediaStore.Images.Media.DATA};
		String result=null;
		CursorLoader cursorLoader=new CursorLoader(context,contentUri,proj,null,null,null);
		Cursor cursor=cursorLoader.loadInBackground();
		if(cursor!=null){
			int column_index=cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA);
			cursor.moveToFirst();
			result=cursor.getString(column_index);
			cursor.close();
		}
		return result;
	}
	public static String getRealPathFromURI_BelowAPI11(Context context,Uri contentUri){
		String[] proj={MediaStore.Images.Media.DATA};
		Cursor cursor=context.getContentResolver().query(contentUri,proj,null,null,null);
		int column_index=0;
		String result="";
		if(cursor!=null){
			column_index=cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA);
			cursor.moveToFirst();
			result=cursor.getString(column_index);
			cursor.close();
			return result;
		}
		return result;
	}
	@SuppressLint("NewApi") public static String getRealPathFromURI_API19(final Context context,final Uri uri){
		if(DocumentsContract.isDocumentUri(context,uri)){
			if(isExternalStorageDocument(uri)){
				final String docId=DocumentsContract.getDocumentId(uri);
				final String[] split=docId.split(":");
				final String type=split[0];
				if("primary".equalsIgnoreCase(type)){
					if(split.length>1){
						return Environment.getExternalStorageDirectory()+"/"+split[1];
					}else{
						return Environment.getExternalStorageDirectory()+"/";
					}
				}else{
					return "storage"+"/"+docId.replace(":","/");
				}
			}else
				if(isDownloadsDocument(uri)){
					String fileName=getFilePath(context,uri);
					if(fileName!=null){
						return Environment.getExternalStorageDirectory().toString()+"/Download/"+fileName;
					}
					final String id=DocumentsContract.getDocumentId(uri);
					final Uri contentUri=ContentUris.withAppendedId(Uri.parse("content://downloads/public_downloads"),Long.valueOf(id));
					return getDataColumn(context,contentUri,null,null);
				}else
					if(isMediaDocument(uri)){
						final String docId=DocumentsContract.getDocumentId(uri);
						final String[] split=docId.split(":");
						final String type=split[0];
						Uri contentUri=null;
						if("image".equals(type)){
							contentUri=MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
						}else
							if("video".equals(type)){
								contentUri=MediaStore.Video.Media.EXTERNAL_CONTENT_URI;
							}else
								if("audio".equals(type)){
									contentUri=MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
								}
						final String selection="_id=?";
						final String[] selectionArgs=new String[]{split[1]};
						return getDataColumn(context,contentUri,selection,selectionArgs);
					}
		}else
			if("content".equalsIgnoreCase(uri.getScheme())){
				if(isGooglePhotosUri(uri))
					return uri.getLastPathSegment();
				return getDataColumn(context,uri,null,null);
			}else
				if("file".equalsIgnoreCase(uri.getScheme())){
					return uri.getPath();
				}
		return null;
	}
	public static String getDataColumn(Context context,Uri uri,String selection,String[] selectionArgs){
		final String column="_data";
		final String[] projection={column};
		try(Cursor cursor=context.getContentResolver().query(uri,projection,selection,selectionArgs,null)){
			if(cursor!=null&&cursor.moveToFirst()){
				final int index=cursor.getColumnIndexOrThrow(column);
				return cursor.getString(index);
			}
		}
		return null;
	}
	public static String getFilePath(Context context,Uri uri){
		final String[] projection={MediaStore.MediaColumns.DISPLAY_NAME};
		try(Cursor cursor=context.getContentResolver().query(uri,projection,null,null,null)){
			if(cursor!=null&&cursor.moveToFirst()){
				final int index=cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DISPLAY_NAME);
				return cursor.getString(index);
			}
		}
		return null;
	}
	public static boolean isExternalStorageDocument(Uri uri){
		return "com.android.externalstorage.documents".equals(uri.getAuthority());
	}
	public static boolean isDownloadsDocument(Uri uri){
		return "com.android.providers.downloads.documents".equals(uri.getAuthority());
	}
	public static boolean isMediaDocument(Uri uri){
		return "com.android.providers.media.documents".equals(uri.getAuthority());
	}
	public static boolean isGooglePhotosUri(Uri uri){
		return "com.google.android.apps.photos.content".equals(uri.getAuthority());
	}
}

File: utilites/src/main/java/com/satya/utilites/utilities/FileUtils.kt
package com.satya.utilites.utilities

import android.annotation.SuppressLint
import android.content.ContentUris
import android.content.Context
import android.net.Uri
import android.os.Environment
import android.provider.DocumentsContract
import android.provider.MediaStore

@SuppressLint("Recycle") object FileUtils {
	fun realPathFromURI(context: Context, uri: Uri): String? {
		when {
			DocumentsContract.isDocumentUri(context, uri) -> when {
				isExternalStorageDocument(uri) -> {
					val docId = DocumentsContract.getDocumentId(uri)
					val split = docId.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
					return if ("primary".equals(split[0], ignoreCase = true)) {
						if (split.size > 1) Environment.getExternalStorageDirectory().toString() + "/" + split[1]
						else Environment.getExternalStorageDirectory().toString() + "/"
					} else "storage" + "/" + docId.replace(":", "/")
				}
				isDownloadsDocument(uri) -> {
					val fileName = getFilePath(context, uri)
					if (fileName != null) return Environment.getExternalStorageDirectory().toString() + "/Download/" + fileName
					return getDataColumn(context, ContentUris.withAppendedId(Uri.parse("content://downloads/public_downloads"), java.lang.Long.valueOf(DocumentsContract.getDocumentId(uri))), null, null)
				}
				isMediaDocument(uri) -> {
					val split = DocumentsContract.getDocumentId(uri).split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
					var contentUri: Uri? = null
					when (split[0]) {
						"image" -> contentUri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI
						"video" -> contentUri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI
						"audio" -> contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
					}
					return getDataColumn(context, contentUri, "_id=?", arrayOf(split[1]))
				}
			}
			"content".equals(uri.scheme!!, ignoreCase = true) -> return if (isGooglePhotosUri(uri)) uri.lastPathSegment else getDataColumn(context, uri, null, null)
			"file".equals(uri.scheme!!, ignoreCase = true) -> return uri.path
		}
		return null
	}
	
	private fun getDataColumn(context: Context, uri: Uri?, selection: String?, selectionArgs: Array<String>?): String? {
		val column = "_data"
		context.contentResolver.query(uri!!, arrayOf(column), selection, selectionArgs, null)!!.use { cursor -> if (cursor.moveToFirst()) return cursor.getString(cursor.getColumnIndexOrThrow(column)) }
		return null
	}
	
	private fun getFilePath(context: Context, uri: Uri): String? {
		context.contentResolver.query(uri, arrayOf(MediaStore.MediaColumns.DISPLAY_NAME), null, null, null)!!.use { if (it.moveToFirst()) return it.getString(it.getColumnIndexOrThrow(MediaStore.MediaColumns.DISPLAY_NAME)) }
		return null
	}
	
	private fun isExternalStorageDocument(uri: Uri): Boolean = "com.android.externalstorage.documents" == uri.authority
	private fun isDownloadsDocument(uri: Uri): Boolean = "com.android.providers.downloads.documents" == uri.authority
	private fun isMediaDocument(uri: Uri): Boolean = "com.android.providers.media.documents" == uri.authority
	private fun isGooglePhotosUri(uri: Uri): Boolean = "com.google.android.apps.photos.content" == uri.authority
}